pkgs/development/compilers/cudatoolkit/redist/find_outputs/find_outputs/outputs.py

```diff
@@ -0,0 +1,311 @@
+        )
+
+    _shared_libs = list(get_dynamic_libraries(store_path))
+    logging.debug(f"Found {len(_shared_libs)} shared libraries.")
+    _shared_lib_dirs = {path.parent.relative_to(store_path) for path in _shared_libs}
+    logging.debug(
+        f"Found {len(_shared_lib_dirs)} shared library directories: {_shared_lib_dirs}"
+    )
+
+    if len(_shared_lib_dirs) > 0 and not any(
+        shared_lib_dir.is_relative_to("lib") for shared_lib_dir in _shared_lib_dirs
+    ):
+        logging.warning(
+            f"Found directories containing shared libraries outside lib: {_shared_lib_dirs}."
+        )
+
+    _static_libs = list(get_static_libraries(store_path))
+    logging.debug(f"Found {len(_static_libs)} static libraries.")
+    _static_lib_dirs = {path.parent.relative_to(store_path) for path in _static_libs}
+    logging.debug(
+        f"Found {len(_static_lib_dirs)} static library directories: {_static_lib_dirs}"
+    )
+
+    if len(_static_lib_dirs) > 0 and not any(
+        static_lib_dir.is_relative_to("lib") for static_lib_dir in _static_lib_dirs
+    ):
+        logging.warning(
+            f"Found directories containing static libraries outside lib: {_static_lib_dirs}."
+        )
+
+    return features
+
+
+def process_package(
+    package_name: str, package: manifest.Package
+) -> tuple[str, Package]:
+    package_info_kwargs: dict[str, str] = {
+        "name": package.name,
+        "license": package.license,
+        "version": package.version,
+    }
+    logging.info(f"Package: {package_name} ({package.name})")
+    logging.debug(f"License: {package.license}")
+    logging.info(f"Version: {package.version}")
+
+    package_arch_kwargs: dict[generic.Architecture, ReleaseFeatures] = {}
+    for arch, release in package.get_architectures().items():
+        if release is None:
+            # Package does not support this architecture.
+            continue
+
+        logging.info(f"Architecture: {arch}")
+        logging.debug(f"Relative path: {release.relative_path}")
+        logging.debug(f"SHA256: {release.sha256}")
+        logging.debug(f"MD5: {release.md5}")
+        logging.debug(f"Size: {release.size}")
+
+        # Get the store path for the release.
+        archive: NixStoreEntry = nix_store_prefetch_file(release)
+        unpacked: NixStoreEntry = nix_store_unpack_archive(archive.store_path)
+        unpacked_root = unpacked.store_path
+        features = get_release_features(unpacked_root)
+        package_arch_kwargs[arch] = features
+
+    return package_name, Package.parse_obj(package_info_kwargs | package_arch_kwargs)
+
+
+def process_manifest(
+    manifest: dict[str, manifest.Package],
+) -> dict[str, Package]:
+    """
+    Processes a manifest to predict the outputs of each package.
+    """
+    with multiprocessing.Pool() as pool:
+        return dict(pool.starmap(process_package, manifest.items(), chunksize=1))
```

Comment:
Is dealing with multiprocessing really necessary here? How long does this script take running serially? (I have had nothing but headaches dealing with concurrency in python)

Reply:
The multi-processing is done mostly to speed up the archive downloading and unpacking. I had it take something like 30m+ the first time I ran it sequentially due to all the downloading and unpacking that ran in serial.

Thankfully, most of the work being done in this script is either handled by Nix (downloading / unpacking archives) or the OS (all the Path utilities).

On subsequent runs, once everything has been downloaded and unpacked, there's no real need for the multiprocessing.

Would you prefer I instead limit the multi-processing to the Nix portion, so only the fetching and unarchiving is done in parallel?